orchestra/utils/task_lifecycle.py

```diff
@@ -275,6 +228,182 @@ def assign_task(worker_id, task_id):
+        if not fake:
+            # Delete assignments created after the revert datetime.
+            assignments_after.delete()
+
+    for assignment in assignment_history(task).all():
+        assignment_audit = _revert_assignment_to_datetime(
+            assignment, revert_datetime, fake)
+        if assignment_audit:
+            audit['assignments'].append(assignment_audit)
+
+    # Delete task if it starts after revert_datetime
+    if task.start_datetime > revert_datetime:
+        if not fake:
+            task.delete()
+        audit['change'] = 'deleted'
+    elif audit['assignments']:
+        if not fake:
+            _revert_task_status(task)
+        audit['change'] = 'reverted'
+
+    return audit
+
+
+@transaction.atomic
+def _revert_assignment_to_datetime(assignment, revert_datetime, fake=False):
+    audit = None
+    # Revert assignment to datetime by removing snapshots
+    gt_idx = None
+    snapshots = assignment.snapshots['snapshots']
+    for i, snapshot in enumerate(snapshots):
+        if parse_datetime(snapshot['datetime']) > revert_datetime:
+            gt_idx = i
+            break
+    if gt_idx is not None:
+        # Revert to before this snapshot
+        audit = {
+            'assignment': assignment,
+            'change': 'reverted',
+            'snapshots': [
+                {
+                    'index': snapshots.index(snapshot),
+                    'snapshot': snapshot,
+                    'change': 'deleted'
+                }
+                for snapshot in snapshots[gt_idx:]
+            ]
+        }
+        if not fake:
+            assignment.snapshots['snapshots'] = snapshots[:gt_idx]
+            assignment.save()
+    return audit
+
+
+@transaction.atomic
+def _revert_task_status(task, fake=False):
+    assignments = assignment_history(task)
+    num_assignments = assignments.count()
+
+    reverted_status = None
+    current_assignment_counter = None
+    previous_assignment_counter = None
+    if num_assignments == 0:
+        # No assignment is present
+        reverted_status = Task.Status.AWAITING_PROCESSING
+    else:
+        last_active_assignment = _get_last_active_assignment(task)
+        if not last_active_assignment:
+            # Task has an assignment but no snapshots
+            reverted_status = Task.Status.PROCESSING
+            current_assignment_counter = 0
+        else:
+            latest_counter = last_active_assignment.assignment_counter
+            snapshot = last_active_assignment.snapshots['snapshots'][-1]
+            if snapshot['type'] == TaskAssignment.SnapshotType.REJECT:
+                # Task was last rejected back to previous worker
+                reverted_status = Task.Status.POST_REVIEW_PROCESSING
+                current_assignment_counter = latest_counter - 1
+                previous_assignment_counter = latest_counter
+
+            elif (snapshot['type'] == TaskAssignment.SnapshotType.SUBMIT
```

Comment:
alternative: `snapshot['type'] in (TaskAssignment.SnapshotType.SUBMIT,TaskAssignment.SnapshotType.ACCEPT)` but go with whichever you feel is better